click_odoo_contrib/update.py

```diff
@@ -97,6 +184,26 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
     Registry.new(database, update_module=True)
     with conn.cursor() as cr:
         _save_installed_checksums(cr)
+
+
+@contextmanager
+def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
+    # Watch for database locks while Odoo updates
+    watcher_process = multiprocessing.Process(
+        target=_db_lock_watcher,
+        args=(database, ctx.params["watcher_beat"], ctx.params["watcher_max_seconds"]),
+    )
+    try:
+        if ctx.params["watcher_max_seconds"] > 0:
+            watcher_process.start()
+        # Update Odoo datatabase
+        _update_db(database, ctx.params["update_all"], ctx.params["i18n_overwrite"])
+    finally:
+        if watcher_process.is_alive():
+            _logger.info("Terminating DB lock watcher")
+            watcher_process.terminate()
+            watcher_process.join()
```

Comment:
The watcher termination is simpler because it should always work.

Reply:
If you start the watcher as thread, there is no need for code to terminate it.

Reply:
I prefer to use processes, to make sure they are not blocked by the GIL. Terminating it is not so hard.

Reply:
Or you can do `watcher_process.daemon = True` so termination will be automatic.

A thread is still better in this case: no risk of leaving processes hanging around and the GIL is not an issue since the watcher mostly does sleeps and psycopg calls which release the GIL anyway.

Reply:
The daemon process is supposed to be reaped automatically when the parent process exits, but the parent process is the current thread, so if we rely in automatic termination, it would leave the watcher working after the update has finished. It's not gonna do too much probably, but I prefer to terminate it in the exact moment when it's not needed anymore. The same applies for auto-closing threads.

About the GIL, I'm more worried the foreground Odoo update process locking the GIL, not actually the watcher.

This watcher is critical for CD/HA, so I prefer to go for the 100% secure way to handle it, even if it means having 4 LOC more. :smile:

Reply:
@Yajo I thought about this a bit more and did some tests with multiprocessing. My conclusion is that it's too easy to have the watcher process continuing after termination of the parent process. In my experiments, a SIGTERM to the parent is enough to leave the watcher running: that's sounds dangerous to me. 

A daemon thread will not have such issue.

In don't think the GIL would cause any issue in this case, as the GIL is relased with any postgress call.

Reply:
OK, I just changed it to thread. I think we should leave this open until I check there's no GIL problem.